### PR TITLE
Exit when CRDs are found for the first time to register with watches

### PR DESCRIPTION
**Pull request by bindrad**
Problem description can be found [here](https://github.com/red-hat-storage/managed-fusion-agent/issues/28)

To build the cache CRDs for the offering should exist in the cluster but Agent installs the offering and then CRDs are created into the cluster. Because of this, the manager was not able to start.

So as a quick fix, we can run the controller initially to just install the offering(without watching offering resources) and then kill the container and re-run it with the offering watches.

### CODE REVIEW

**Comment by nb-ohad** [comment]
Please add to the description a link to the github issue capturing the problem

File: controllers/managedfusionoffering_controller.go [hunk 1]
```diff
@@ -65,11 +72,42 @@ func (r *ManagedFusionOfferingReconciler) SetupWithManager(mgr ctrl.Manager, ctr
 		Owns(&opv1a1.CatalogSource{}).
 		Owns(&opv1a1.Subscription{})
 
-	pluginSetupWatches(controllerBuilder)
+	if doesCRDsExists() {
+		r.DoesCRDExistsAtStart = true
+		pluginSetupWatches(controllerBuilder)
+	}
```
Comment (nb-ohad) [inline controllers/managedfusionoffering_controller.go:78]: We have an established way to check for CRD existence. It was added with the SDN/OVN contribution a couple of months back on the original deployer repo. Please use this

https://github.com/red-hat-storage/ocs-osd-deployer/blob/main/controllers/managedocs_controller.go#L323-L328


Also, why is this implemented at the offering reconciler level where it is a fix for df add watches? This should go into dfSetupWatches. This all hack should be internal to the DF offering.

File: controllers/managedfusionoffering_controller.go [hunk 2]
```diff
@@ -65,11 +72,42 @@ func (r *ManagedFusionOfferingReconciler) SetupWithManager(mgr ctrl.Manager, ctr
...
+		if errors.IsNotFound(err) {
+			return false
+		} else {
+			setupLog.Error(err, fmt.Sprintf("Error Getting CRD for %s", "storageclusters.ocs.openshift.io"))
+			os.Exit(1)
+		}
+	}
+
+	return true
+}
```
Comment (nb-ohad) [inline controllers/managedfusionoffering_controller.go:109]: As of the last comment
```suggestion
```

File: controllers/managedfusionoffering_controller.go [hunk 3]
```diff
@@ -84,11 +122,20 @@ func (r *ManagedFusionOfferingReconciler) Reconcile(ctx context.Context, req ctr
...
 	} else {
-		return result, nil
+		if result, err := pluginReconcile(r, r.managedFusionOffering); err != nil {
+			return ctrl.Result{}, fmt.Errorf("An error was encountered during reconcile: %v", err)
+		} else {
+			return result, nil
+		}
 	}
+
+	return ctrl.Result{}, nil
```
Comment (nb-ohad) [inline controllers/managedfusionoffering_controller.go:138]: All of this code is very specific for DF, even if it does not say DF in the variable names, we should not do that. For this workaround, only the actual offering should know if it needs a restart or not and that code should go into the dfReconcile code at the start. But the reconciler should implement an `r.Exist()` function to allow the offering to exit without using a hack.  
```suggestion
	if result, err := pluginReconcile(r, r.managedFusionOffering); err != nil {
		return ctrl.Result{}, fmt.Errorf("An error was encountered during reconcile: %v", err)
	} else {
		return result, nil
	}
```

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: controllers/managedfusionoffering_controller.go [hunk 4]
```diff
@@ -51,6 +57,7 @@ type ManagedFusionOfferingReconciler struct {
 	operatorGroup         *opv1.OperatorGroup
 	catalogSource         *opv1a1.CatalogSource
 	subscription          *opv1a1.Subscription
+	DoesCRDExistsAtStart  bool
```
[the remark this answers is not in the document]
Comment (nb-ohad) [reply controllers/managedfusionoffering_controller.go:60]: @leelavg You cannot wait in an init container for something that the same entity needs to initiate the installation for

**Comment by bindrad** [comment]
https://github.com/red-hat-storage/managed-fusion-agent/pull/35 implements the workaround so closing this PR

**Comment by leelavg** [comment]
should be closed in favour of #54?

**Comment by bindrad** [comment]
closed in favour of https://github.com/red-hat-storage/managed-fusion-agent/pull/54

**Pull request closed without merging**